Particle Systems

jasonbeverage opened the pull request:
Me again 😄 

This PR adds particle system support to Cesium.  This work is a conglomeration of ideas from:
-  The original particles branch that @bagnell did
-  osgParticle from OpenSceneGraph
-  Unity's particle system
-  My own ideas :)

I'm sure you'll have plenty of feedback on this one so I wanted to get it out early.

I've got a live demo you can play with here:
http://cesiumparticles.netlify.com/apps/sandcastle/?src=ParticleSystem.html&label=Showcases

Originally I was going to write a fully GPU based particle system (I hacked one together using OpenSceneGraph, see this [video](https://youtu.be/17tUEsjQ9Eg))  but ultimately just ended up doing it all on the CPU.  The GPU solution is really interesting for when you need to render a massive amount of particles for something like a water or super realistic fire effect, but it also complicates integrating logic that doesn't lend itself well to being fully implemented on the GPU.  For example, in my early work on this I was able to integrate the [Cannon.js](http://www.cannonjs.org/)  rigid body engine into Cesium with pretty minimal code which would have been really difficult to do on the GPU.  You can see a prototype video [here](https://youtu.be/-VBmaY_6yDY)

Some notes about the design
- A ParticleSystem has an emitter.  The emitter defines the initial position of a particle as well as it's initial direction.  This is similar to Unity's "shape" property.  I've implemented a few emitters, each with their own properties defining the shape.  You can select various emitter shapes in the sand castle demo from the drop down menu.
  - BoxEmitter
  - CircleEmitter
  - ConeEmitter
  - SphereEmitter
- The ParticleSystem has it's own modelMatrix, making it easy to attach to an entity through the Entity framework.  This only effects the initial position and direction of the particle.  Once a particle is emitted it is independent of that modelMatrix.  For example, you can click the "fly" checkbox in the demo and the plane will fly, emitting particles along the way.  But each particle will still maintain it's own frame of reference.
-  ParticleSystem also has an emitterModelMatrix.  This is an additional transform that occurs within the localized space of the modelMatrix.  For example, it is how I'm positioning the emitter on the propeller of the plane instead of the center of the model.  If you modify the translation or rotation parameters in the UI it will change the position of the emitter within the model.
-  The number of particles emitted is controlled by the rate parameter, which is simply the number of particles to to emit each second.  You'll also notice an explosion every 5 seconds in the demo.  That is controlled by the "burst" array (which is an idea I took from Unity).  You can define timed bursts of particles like this:

```
bursts: [
            {time: 5.0, min: 300, max: 500},
            {time: 10.0, min: 50, max: 100},
            {time: 15.0, min: 200, max: 300}
        ]
```
The time is the amount of time into the particle system simulation loop before the burst is applied and it will emit somewhere between min and max particles on the burst.  ParticleSystem also has a loop option, which is mostly just used to reset the bursts when it's done looping.
-  The lifetime property of a particle system defines how long the system is active.  For example, if you set it to 10 particles will emit for 10 seconds and then stop.  By default it's infinite.
-  Properties that define the initial state of the particle that can vary are prefixed with min/max.  For example, min/maxSpeed, min/maxLife, etc.  Properties will be a random value between the min and max.
-  The scale and color of a particle can change over the lifetime of a particle.  These can be set with start/endColor and start/endScale.  This lets you do things like make the particles get smaller as they die, or become more transparent so they fade out.

Things I know that still need to be done:
- [x] Update CHANGES.md
- [x] Unit tests
- [x] ~~Wiring up the "complete" event on particle systems through the entity framework so you can script particle systems.~~
- [x] replace fire.png, https://github.com/AnalyticalGraphicsInc/cesium/pull/5212#issuecomment-296399516 
- [x] Any feedback you have :)

![particlesystem](https://cloud.githubusercontent.com/assets/326626/25094628/10671aec-2366-11e7-9abc-9e1f48989886.jpg)

Thanks!

Jason

pjcozzi commented:
Amazing, thanks @jasonbeverage!  We'll look at this soon with great excitement!

pjcozzi commented:
Part of #3001.

pjcozzi commented:
@bagnell can we delete this branch now: https://github.com/AnalyticalGraphicsInc/cesium/compare/particles.  If so, please go for it.

mramato commented:
Yet another awesome contribution, thanks @jasonbeverage!  The amount of thought and details you put into your pull request descriptions alone should serve as an example for all other contributions.

@rahwang once this gets in, maybe some carefully placed fountains and smoke stacks can be added to the arcade game. 😄

rahwang commented:
Yes! Looks awesome, @jasonbeverage !

jasonbeverage commented:
Thanks for the kind words @mramato .  As an open source maintainer myself I know the value of a good PR :)

jasonbeverage commented:
I just saw your video of the drone racing game, amazing!

pjcozzi commented:
>  Apps/SampleData/fire.png

@jasonbeverage did you create this image or does it need to be credited in LICENSE.md?

Also does it need so much of a transparent border?

pjcozzi commented:
Really nice job with the Sandcastle example; it is too much fun.

pjcozzi commented:
@rahwang could you test this feature out and provide any input?  Perhaps create a new Sandcastle example showing up to show the particle system for a **simple** smoke simulation?

pjcozzi commented:
@jasonbeverage you might like our four-year-old notes on particle systems if you haven't already seen them: https://github.com/AnalyticalGraphicsInc/cesium/wiki/Particle-System-Details

jasonbeverage commented:
Hey @pjcozzi that fire.png is just something I found online.  I have zero art skills, so if someone wants to replace it with something else that would probably be better anyway :)

The sandcastle example is lots of fun, I've wasted too much time playing with it.

pjcozzi commented:
> The ParticleSystem has it's own modelMatrix...
ParticleSystem also has an emitterModelMatrix...

Since there is a one-to-one relationship between a particle system and the emitter, are you sure both of these are required?  Is this an ease-of-use feature for the entity API?  Can you show an example of how both would be used in the same app?

pjcozzi commented:
> Hey @pjcozzi that fire.png is just something I found online. I have zero art skills, so if someone wants to replace it with something else that would probably be better anyway :)

Will do.

@rahwang @austinEng @hpinkos @likangning93 can one of you artistically-inclined folks create a replacement for [fire.png](https://github.com/jasonbeverage/cesium/blob/1a33efb9bdacdf6396170b6677c7b072132a1a94/Apps/SampleData/fire.png) or find one that is CC licensed?

rahwang commented:
@pjcozzi, I'd be happy to do that!

pjcozzi commented on Source/Scene/Particle.js:
```diff
@@ -0,0 +1,82 @@
+    Particle.prototype.update = function(forces, dt) {
+
+        // Apply the velocity
+        var delta = new Cartesian3();
```
Can this be a scratch variable?

jasonbeverage replied:
Yup, made the change.

pjcozzi commented on Source/Scene/Particle.js:
```diff
@@ -0,0 +1,82 @@
+        for (var i = 0; i < length; ++i) {
+            var force = forces[i];
+
+            if (typeof force === 'function') {
```
Can we just require that it is always a simple callback function?  Cesium does this elsewhere.

jasonbeverage replied:
One of the reasons I played with making Force an object in addition to a callback was so that forces could easily maintain their own state and be changeable.  I don't have a concrete example now but I thought about things like making gravity in the sandcastle example a class instead of a callback.

Something like:
```
var gravity = new LocalizedAcceleration(new Cartesian3(0.0, 0.0, -9.8));
forces = [gravity];
```

Then you could change it with:
```
gravity.acceleration = new Cartesian3(...);
```

pjcozzi replied:
I see the potential, but there may be other ways to achieve this and can we just keep this trim and simple until we have concrete use cases, and just roadmap this, #5212?

pjcozzi commented on Source/Scene/Particle.js:
```diff
@@ -0,0 +1,82 @@
+        // Compute the normalized age.
+        if (this.life === Number.MAX_VALUE) {
+            this.normalizedAge = 0.0;
+        }
```
"} else {" on the same line.

pjcozzi replied:
Same comment throughout.

jasonbeverage replied:
Fixed.

jasonbeverage commented:
The modelMatrix just a member of the ParticleSystemPrimitive and is just used to tie the primitive to the entity, so you don't see that used in the sandcastle app, it's used under the hood in the ParticleSystemVisualizer 
https://github.com/jasonbeverage/cesium/blob/particles/Source/DataSources/ParticleSystemVisualizer.js#L103-L107

If you were to use the ParticleSystemPrimitive directly you'd want to set it yourself though.

The emitterModelMatrix is really an ease of use thing is just a local matrix once the particle system is positioned in entity space.  For example, in the sandcastle example, it's easy to position the fire right on the propeller without knowing where the entity is in space.  Or you could rotate it so the fire shoots out of the back of the plane, and it would always do that no matter what direction the plane was flying.

pjcozzi commented on Source/Scene/Particle.js:
```diff
@@ -0,0 +1,82 @@
+            this.normalizedAge = this.age / this.life;
+        }
+
+        // If this particle is older than it's lifespan then die.
```
Perhaps just
```javascript
return this.age <= this.life;
```

jasonbeverage replied:
Fixed.

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+            this.maxSpeed = defaultValue(options.maxSpeed, 1.0);
+        }
+
+        var life = defaultValue(options.life, undefined);
```
Just
```javascript
var life = options.life;
```

pjcozzi replied:
Similar comments throughout this constructor function.

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        this.endScale = defaultValue(options.endScale, 1.0);
+
+        this.rate = defaultValue(options.rate, 5);
+        this.bursts = defaultValue(options.bursts, null);
```
Just
```
this.bursts = options.bursts;
```

jasonbeverage replied:
I made this change along with a few others in acd31e71824a8f0bd75a07432a51091c01d5ef28 that removed the unnecessary use of defaultValue.

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        system._billboardCollection.remove(particle._billboard);
+    }
+
+    function random(a, b) {
```
This is duplicate in a few places.  Should it be part of `CesiumMath`?

jasonbeverage replied:
Made a function for this called randomBetween in CesiumMath.

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        billboard.position = particle.position;
+
+        // Update the color
+        var r = particle.startColor.red + particle.normalizedAge * (particle.endColor.red - particle.startColor.red);
```
Would [lerp](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Math.js#L311) work for these?

jasonbeverage replied:
lerp works great!

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        this.particles.push(particle);
+    };
+
+    ParticleSystem.prototype.calcNumberToEmit = function(dt) {
```
If this would be part of the public API, spell out `calc`.

jasonbeverage replied:
I'll rename it, but also mark it as @private.  It's not something that an external user should be calling.

jasonbeverage replied:
I actually ended up making these local file scoped functions like mentioned [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/CodingGuide/README.md#private-functions)

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+
+        // Apply any bursts
+        if (this.bursts) {
+            for (i = 0; i < this.bursts.length; i++) {
```
Generally, we would assign this outside of the loop
```javascript
var length = this.bursts.length;
```

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        if (this.bursts) {
+            for (i = 0; i < this.bursts.length; i++) {
+                var burst = this.bursts[i];
+                if ((!defined(burst, "complete") || !burst.complete) && this.currentTime > burst.time) {
```
Remove `, "complete"`, right?

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        if (numToEmit > 0 && emitter) {
+
+            // Compute the final model matrix by combining the particle systems model matrix and the emitter matrix.
+            var combinedMatrix = new Matrix4();
```
Worth a scratch?

jasonbeverage replied:
This and the tmp variable are converted to scratch vars in 8a8626cd61e26e2f1f19bd60c1bb6866b613d1a4

pjcozzi commented:
@jasonbeverage have you tried turning off all the graphics (globe, skybox, etc.), using a 1x1 texture, and a ton of particles, and then doing a quick profile to see if anything scary comes up and if there is low hanging fruit for us to optimize?

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+
+            for (i = 0; i < numToEmit; i++) {
+                // Create a new particle.
+                particle = this.emitter.emit( particle );
```
Throughout, just `(particle)` for function calls.

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+                if (particle) {
+
+                    //For the velocity we need to add it to the original position and then multiply by point.
+                    var tmp = new Cartesian3();
```
Can we remove this per-emission allocations?

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+            if (this.loop) {
+                this.currentTime = this.currentTime - this.lifeTime;
+                // Reset any bursts
+                for (i = 0; i < this.bursts.length; i++) {
```
Same `length` comment.

jasonbeverage replied:
Fixed in 7e2a965cd2d59a5b2c4b98196db4ef06f001c106

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+    var ParticleSystem = function(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        this.particles = defaultValue(options.particles, []);
```
Should this property be public?

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,295 @@
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        this.particles = defaultValue(options.particles, []);
+        this.forces = defaultValue(options.forces, []);
```
Should all these properties be get/set?

Likewise, what properties should have a `DeveloperError` thrown for clearly wrong values like negative scale?

pjcozzi commented:
> The emitterModelMatrix is really an ease of use thing is just a local matrix once the particle system is positioned in entity space...

Gotcha, thanks!

jasonbeverage commented:
Hey @pjcozzi I'm going to work on addressing your comments soon.  You mentioned the "autoformatter", is that an npm task I can run or something different?

pjcozzi commented:
> Hey @pjcozzi I'm going to work on addressing your comments soon.

Awesome!

> You mentioned the "autoformatter", is that an npm task I can run or something different?

No, but that would be great!

We use an autoformatter in WebStorm:

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/WebStormGuide/README.md

There is also Ctrl-Shift-F in Eclipse.

I think both are _close_ to our code conventions.

a735561656 commented:
hello pjcozzi，your example of particle is so cool. I'm trying to get the smoking effect that base on yours.I change the image of fire to another,but there is a 'boom effect' once in a while in your example . how to remove it ?

jasonbeverage commented:
There is a bursts property on the particle system that you can remove.
Looking forward to seeing what you do with the particle systems!

On Fri, May 5, 2017, 5:23 AM a735561656 <notifications@github.com> wrote:

> hello pjcozzi，your example of particle is so cool. I'm trying to get the
> smoking effect that base on yours.I change the image of fire to another,but
> there is a 'boom effect' once in a while in your example . how to remove it
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AnalyticalGraphicsInc/cesium/pull/5212#issuecomment-299420207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAT74ofezMwt5J5JqB4nKuQ9tCBv9KM4ks5r2uqBgaJpZM4M_W1S>
> .
>

jasonbeverage commented on Apps/Sandcastle/gallery/ParticleSystem.html:
```diff
@@ -0,0 +1,445 @@
+            R <input type="text" size="5" data-bind="value: roll">
+        </td>
+    </tr>
+
```
Done.

jasonbeverage commented on Apps/Sandcastle/gallery/ParticleSystem.html:
```diff
@@ -0,0 +1,445 @@
+    },
+
+    particleSystem: {
+        image: '../../SampleData/fire.png',
```
Done.

jasonbeverage commented on Apps/Sandcastle/gallery/ParticleSystem.html:
```diff
@@ -0,0 +1,445 @@
+        startScale: viewModel.startScale,
+        endScale: viewModel.endScale,
+
+        minLife: viewModel.minLife,
```
Done.

jasonbeverage commented on Apps/Sandcastle/gallery/ParticleSystem.html:
```diff
@@ -0,0 +1,445 @@
+
+    // Spin the emitter if enabled.
+    if (viewModel.spin) {
+        viewModel.heading += 1;
```
Done.

jasonbeverage commented on Apps/Sandcastle/gallery/ParticleSystem.html:
```diff
@@ -0,0 +1,445 @@
+        }
+        else {
+            entity.position = staticPosition;
+            entity.orientation = null;
```
undefined is just fine, made the change.

jasonbeverage commented on Source/DataSources/ParticleSystemGraphics.js:
```diff
@@ -0,0 +1,331 @@
+         * @type {Property}
+         * @default new CircleEmitter({radius: 0.5});
+         */
+        emitter : createPropertyDescriptor('emitter'),
```
For now, yes, they would just create multiple particle systems.  I played around with ParticleSystems having multiple emitters, and it worked just fine.  It was hard to really distinguish what would be the difference between multiple emitters and multiple systems, so I just tried to keep it simple for now.

jasonbeverage commented on Source/DataSources/ParticleSystemGraphics.js:
```diff
@@ -0,0 +1,331 @@
+         * @memberof ParticleSystemGraphics.prototype
+         * @type {Property}
+         */
+        image : createPropertyDescriptor('image'),
```
I wanted to keep the initial PR simple, but there are definite use cases where you would want particles to change their billboards over time.  For example, osgParticle has a fire example that uses an animated fire particle that just has a bunch of frames in a loop.  So I'd say that is definitely something that would be useful at some point and we should keep in mind.

pjcozzi replied:
Gotcha.  I added this idea to a new roadmap for future work.  I'm sure this is a fun area of the code that folks will want to contribute to.  Please add more ideas to #5314.

jasonbeverage commented on Source/DataSources/ParticleSystemGraphics.js:
```diff
@@ -0,0 +1,331 @@
+        image : createPropertyDescriptor('image'),
+
+        /**
+         * The scale of a particle at the beginning of it's lifetime.
```
Fixed.

jasonbeverage commented on Source/DataSources/ParticleSystemGraphics.js:
```diff
@@ -0,0 +1,331 @@
+        rate: createPropertyDescriptor('rate'),
+
+        /**
+         * The minimum width of a particle billboard.
```
Done.

jasonbeverage commented on Source/DataSources/ParticleSystemGraphics.js:
```diff
@@ -0,0 +1,331 @@
+        loop: createPropertyDescriptor('loop'),
+
+        /**
+         * The lifetime of the particle system.
```
Done.

jasonbeverage commented on Source/DataSources/ParticleSystemVisualizer.js:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    /**
+     * Updates models created this visualizer to match their
```
Copy and pasted the ModelVisualizer to make the ParticleSystemVisualizer and didn't update all the comments ;)  Fixed now.

jasonbeverage commented:
Hey @pjcozzi I pushed a bunch of changes to address your review.  Not quite done, will probably finish the rest tomorrow!

jasonbeverage commented on Source/Scene/Particle.js:
```diff
@@ -0,0 +1,82 @@
+        this.position = Cartesian3.clone(defaultValue(options.position, Cartesian3.ZERO));
+        this.velocity = Cartesian3.clone(defaultValue(options.velocity, Cartesian3.ZERO));
+        this.life = defaultValue(options.life, Number.MAX_VALUE);
+        this.image = defaultValue(options.image, null);
```
Fixed

jasonbeverage commented on Source/DataSources/ParticleSystemGraphics.js:
```diff
@@ -0,0 +1,331 @@
+        },
+
+        /**
+         * Gets or sets the boolean Property specifying the visibility of the model.
```
Fixed, copy and paste strikes again.

jasonbeverage commented on Source/Scene/BoxEmitter.js:
```diff
@@ -0,0 +1,45 @@
+        Particle) {
+    "use strict";
+
+    var BoxEmitter = function(options) {
```
I've changed all of the constructors, I think I copied the "var" syntax from one of the enum classes.

The BoxEmitter just emits particles from a box with width, height and depth parameters.  I don't have a particular use case for a box emitter vs  the other emitter types, but it's apparently useful b/c it's implemented in Unity as a shape type :smile:.  

Playing around in unity you can generate some neat effects by setting one or more of the dimensions of the box to 0 so you could generate particles from a flat plane or a line.

pjcozzi replied:
Gotcha, sounds good!

jasonbeverage commented on Source/Scene/ConeEmitter.js:
```diff
@@ -0,0 +1,51 @@
+        Particle) {
+    "use strict";
+
+    var ConeEmitter = function(options) {
```
Added ParticleEmitter interface class in 76d6ccf266ffeb984f9621d113eba55457017ae1

jasonbeverage commented on Source/Scene/BoxEmitter.js:
```diff
@@ -0,0 +1,45 @@
+/*global define*/
+define([
```
I pushed docs for the emitter classes in 5b95fa232ec545adc88884fd069b864d172ce971 since they should be public.

The plumbing for the entity framework really did seem like lots of duplicate work, but I tried to keep it similar to the way the other visualizers worked.  Basically ParticleSystemGraphics is just storing properties for the ParticleSystem primitive.  The only interesting thing the visualizer does is ties the entity's model matrix to the particle system primitive's model matrix.  

Originally I just had the sandcastle demo using the ParticleSystem primitive directly and just tying it to the entity in a preFrame callback and the code was pretty straightforward, although not as obvious as the entity framework.

If you can think of a way to slim down the ParticleSystemGraphics I'd be all for it, I was just trying to keep it similar to the other things in the entity framework.

pjcozzi replied:
My gut was that this would just be a primitive - which almost cuts the code/doc/tests in half.  If we want to put this in CZML later then we may want the Cesium `Property` system with the entity layer.

We plan to streamline the entity API eventually.  Perhaps it would be best to park the entity API for this in a branch until we have that figured out?  It will be much less to maintain.

pjcozzi replied:
@bagnell I would still do this as a primitive; there's still too much plumbing and two APIs here for basically the ability to avoid setting a model matrix and consistency in one sense (when used with entities) and inconsistency in other sense (when used with 3D Tiles).  We could add CZML later if desired.

jasonbeverage commented on Source/Scene/Particle.js:
```diff
@@ -0,0 +1,82 @@
+        this.age = 0.0;
+        this.normalizedAge = 0.0;
+
+        this.startColor = defaultValue(options.startColor, Color.clone(Color.WHITE));
```
Fixed in f2b0a7bdef9096936b1880c4bc1dd07f3c4d359d

jasonbeverage commented:
Hey @pjcozzi, I pushed a change in 319b9ce that uses a pool of particles to avoid allocating so many of them.

jasonbeverage commented:
At this point I'd like to leave anything left with getting this merged to you if that's ok, I'm strapped for time right now and getting pulled away for more work :)

pjcozzi commented:
> At this point I'd like to leave anything left with getting this merged to you if that's ok

We're happy to.  We have a code sprint next month so we should be able to get this into the 1.35 release in July.

Thanks again, this will be hugely valuable.

bagnell commented:
@pjcozzi This is ready for another look.

All of the comments have been addressed, doc updated and GC minimized. From a profile (only a particle system in the scene), the GC was less than 0.1%. Most of the time spent is in the `BillboardCollection.update` and `bufferSubData` (also setting uniforms) which is what I was expecting.

What remains are tests and a new particle image.

pjcozzi commented:
@rahwang do you still plan to make the new particle image?  We'd like to merge this soon.

pjcozzi commented on Apps/Sandcastle/gallery/ParticleSystem.html:
```diff
@@ -0,0 +1,445 @@
+
+var gravityScratch = new Cesium.Cartesian3();
+
+function applyGravity(p, dt) {
```
Bump @bagnell

bagnell replied:
I wouldn't make it part of Cesium. It is here and as a code example for the force callbacks.

pjcozzi replied:
OK

pjcozzi commented:
Update CHANGES.md

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,539 @@
+        CircleEmitter) {
+    "use strict";
+
+
```
Whitespace

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,539 @@
+         * @type {Number}
+         * @default 1.0
+         */
+        this.startScale = defaultValue(options.startScale, 1.0);
```
What are the bounds on all these?  Do any of these need explicit `set` functions to throw `DeveloperError`s?

pjcozzi commented on Source/Scene/ParticleSystem.js:
```diff
@@ -0,0 +1,539 @@
+    /**
+     * Gets a Particle to add to the particle system, reusing Particles from the pool if possible.
+     */
+    function getOrCreateParticle() {
```
Are you sure memory management this simple is OK?  E.g., we don't need to time stamp particles in `particlePool` and free over time?

Have you also tried multiple particle systems in the same scene?

bagnell replied:
This is tricky. It depends on the rate particles are emitted, the min/max life of the particles, and bursts of emitted particles.

I updated the update to free the particles/billboards every x number of frames. We should revisit at a later time.

pjcozzi replied:
If the profile looks good, keep it simple for now.  Maybe a `PERFORMANCE_IDEA` if you want.

bagnell replied:
Updated #5314.

bagnell commented:
@pjcozzi This is ready for another review. Since your last review, I added tests, updated CHANGES.md, and addressed all of your comments. I'm working on an additional Sandcastle example, but that can be a separate PR.

bagnell commented:
I added better memory management so its not adding and removing billboards causing constant VAO creation/destruction. Also, added a new Sandcastle example.

pjcozzi commented:
> Apps/SampleData/fire.png

Are we sure this should have so much alphaed blank space around it?  That is a lot of wasted fillrate.

pjcozzi commented:
In the Sandcastle example, the "Show" button turns off the aircraft but not the particle system.

pjcozzi commented:
In the fireworks Sandcastle example, I would not expect this to be 10% of the profile?  Is this an easy fix for this PR or should we submit an issue?

![image](https://user-images.githubusercontent.com/782098/27407958-802e2abc-56a8-11e7-89ca-9b3ef1a8a5e1.png)

pjcozzi commented:
Just those comments.  Very close!

bagnell commented:
> In the fireworks Sandcastle example, I would not expect this to be 10% of the profile? Is this an easy fix for this PR or should we submit an issue?

I would submit an issue. The problem is the memory management again. Sometimes more particles are created than was estimated and they get removed later. This causes the number of commands from the billboard collection to change. For every new command, it generates the OIT shader and looks it up in the cache.

bagnell commented:
@pjcozzi This is ready for another look.

pjcozzi commented:
Submitted #5522

pjcozzi commented:
Thanks again @jasonbeverage @bagnell!

@jasonbeverage what's next?!?!? 😃

rahwang commented:
This is so cool!

jasonbeverage commented:
Thanks again for working with me to get this merged!  I'll hopefully have a couple more PRs before the end of the year ;)

pjcozzi commented:
Really looking forward to them, any teasers???

jasonbeverage commented:
No teasers yet, still working on what they will actually be :)
